File: Valid Number.java
// Leetcode Problem 65: Valid Number
// JAVA CODE
class Solution {
    public boolean isNumber(String s) {
        s = s.trim();
        return s.matches("[+-]?((\\d+(\\.\\d*)?)|(\\.\\d+))([eE][+-]?\\d+)?");
    }
}

File: Valid Number.py
# Leetcode Problem 65: Valid Number
# PYTHON CODE
class Solution:
    def isNumber(self, s: str) -> bool:
        s = s.strip()

        seen_digit = False
        seen_dot = False
        seen_exp = False

        for i, ch in enumerate(s):
            if ch.isdigit():
                seen_digit = True

            elif ch in ['+', '-']:
                if i > 0 and s[i - 1] not in ['e', 'E']:
                    return False

            elif ch == '.':
                if seen_dot or seen_exp:
                    return False
                seen_dot = True

            elif ch in ['e', 'E']:
                if seen_exp or not seen_digit:
                    return False
                seen_exp = True
                seen_digit = False

            else:
                return False

        return seen_digit